exporters/otlp/trace/src/main/java/io/opentelemetry/exporter/otlp/trace/InstrumentationLibraryMarshaler.java

```diff
@@ -6,31 +6,49 @@
 package io.opentelemetry.exporter.otlp.trace;
 
 import com.google.protobuf.CodedOutputStream;
+import io.opentelemetry.context.internal.shaded.WeakConcurrentMap;
 import io.opentelemetry.proto.common.v1.InstrumentationLibrary;
 import io.opentelemetry.sdk.common.InstrumentationLibraryInfo;
 import java.io.IOException;
+import java.io.UncheckedIOException;
 
 final class InstrumentationLibraryMarshaler extends MarshalerWithSize {
-  private final byte[] name;
-  private final byte[] version;
 
-  static InstrumentationLibraryMarshaler create(InstrumentationLibraryInfo libraryInfo) {
-    byte[] name = MarshalerUtil.toBytes(libraryInfo.getName());
-    byte[] version = MarshalerUtil.toBytes(libraryInfo.getVersion());
+  private static final WeakConcurrentMap<
+          InstrumentationLibraryInfo, InstrumentationLibraryMarshaler>
+      LIBRARY_MARSHALER_CACHE = new WeakConcurrentMap.WithInlinedExpunction<>();
+
+  private final byte[] serializedInfo;
 
-    return new InstrumentationLibraryMarshaler(name, version);
+  static InstrumentationLibraryMarshaler create(InstrumentationLibraryInfo libraryInfo) {
+    InstrumentationLibraryMarshaler cached = LIBRARY_MARSHALER_CACHE.get(libraryInfo);
+    if (cached == null) {
+      // Since WeakConcurrentMap doesn't support computeIfAbsent, we may end up doing the conversion
+      // a few times until the cache gets filled which is fine.
```

Comment:
I don't think this is even an issue, since the exporter that will be doing this marshalling is single-threaded. :)